src/MerkleTreeBuilder.sol

```diff
@@ -0,0 +1,193 @@
+// SPDX-License-Identifier: GPL-3.0
+pragma solidity 0.8.26;
+
+import { IRegistrarLike } from "./interfaces/IRegistrarLike.sol";
+import { LinkedList, SortedLinkedList } from "./libs/SortedLinkedList.sol";
+
+contract MerkleTreeBuilder {
+    using SortedLinkedList for LinkedList;
+
+    /* ========== ERRORS ========== */
+
+    error ListAlreadyExists();
+    error InvalidList();
+    error InvalidAdd();
+    error InvalidRemove();
+    error NotInList();
+    error ValueInList();
+    error ValueNotInList();
+
+    /* ========== EVENTS ========== */
+
+    event RootUpdated(bytes32 indexed list, bytes32 root);
+
+    /* ========== STATE ========== */
+
+    uint8 internal constant ZERO_BIT = 0;
+    uint8 internal constant ONE_BIT = 1;
+    address public immutable registrar;
+    bytes32 internal constant ZERO_WORD = bytes32(0);
+    mapping(bytes32 => LinkedList) internal _lists;
+    mapping(bytes32 => bytes32) internal _roots;
+
+    /* ========== CONSTRUCTOR ========== */
+
+    constructor(address registrar_) {
+        registrar = registrar_;
+    }
+
+    /* ========== MANAGE LISTS ========== */
+
+    function addToList(bytes32 list, bytes32 before, bytes32 value) external {
+        // Check that the value is set on the list in the registrar
+        if (!_isSetOnRegistrar(list, value)) revert InvalidAdd();
```

Comment:
opposite check? 
```suggestion
        if (_isSetOnRegistrar(list, value)) revert InvalidAdd();
```

Reply:
this merkle tree should only contain values that exist in the registrar so the check is correct